### DISCUSSION TITLE
merge_noun_chunks: + replace named entities

### DISCUSSION THREAD

**Discussion by simonschoe** [explosion/spaCy#8646] [Help: Coding & Implementations] [answered]
Hi there,
I am struggling with the replacement of named entities by their labels in merged noun chunks. For example, I would like the `noun_chunk` `more than 1,000,000 subscribers` to be returned as `more than NUM subscribers`. On the token-level that is not an issue at all, but since the `merge_noun_chunk` pipe seems to retokenize when calling `pipe()`, I have no idea how to do the replacement afterwards. Here my code:

```python
import spacy

nlp = spacy.load(
    'en_core_web_trf',
    disable=[],
    config={'components.lemmatizer.mode': 'rule'}
)
nlp.add_pipe('merge_noun_chunks')

t = ["We are the largest online entertainment subscription service in the United States providing more than 1,000,000 subscribers access to a comprehensive library of more than 14,500 movie, television and other filmed entertainment titles."]

docs = nlp.pipe(t, n_process=1, batch_size=2)

for doc in docs:
    
    tokenized = [token.lemma_ if not token.ent_type_ else token.ent_type_ for token in doc if all([token.is_stop == False, token.like_num == False, token.is_punct == False, len(token.lemma_) > 1])]

tokenized
```
```python
['the large online entertainment subscription service',
 'GPE',
 'provide',
 'more than 1,000,000 subscriber',
 'access',
 'a comprehensive library',
 'more than 14,500 movie, television and other film entertainment title']
```

**Comment by polm** [comment, marked as answer]
This is going to be pretty tricky if you want to preserve the Doc structure. spaCy docs are designed to preserve the input string without modification, so rewriting the token and then retokenizing directly isn't an option.

What you can do is get the attributes you care about, and then use [`Doc.__init__`](https://spacy.io/api/doc#init) to construct a doc with all the same attributes but just some surface text changed. If you first do that with a doc that didn't use `merge_noun_chunks`, you can call the merge as a separate function afterwards. 

Here's some example code:

```
import spacy
from spacy.pipeline.functions import merge_noun_chunks
from spacy.tokens import Doc

nlp = spacy.load("en_core_web_sm")

text = "I have 17 apples"
doc = nlp(text)
for chunk in doc.noun_chunks:
    print(chunk)

print("-----")

def hide_num(tok):
    return "NUM" if tok.like_num else tok.text

pos = [tok.pos_ for tok in doc]
spaces = [bool(tok.whitespace_) for tok in doc]
words = [hide_num(tok) for tok in doc]
deps = [tok.dep_ for tok in doc]
heads = [tok.head.i for tok in doc]

doc2 = Doc(nlp.vocab, words=words, spaces=spaces, pos=pos, deps=deps, heads=heads)

for chunk in merge_noun_chunks(doc2).noun_chunks:
    print(chunk)
```

**Reply by simonschoe** [reply]
Hi @polm thanks for the reply and help! I must admit, initially I had hoped there would be an elegant solution where I could simply iterate over each token in the noun_chunks in kind of a sub-loop... But as far as I concerned, this approach might work.

However, what I couldn't quite figure out yet: Where in the workflow would I remove stop words or punctuation? In particular I would like to have that removed on the token-level but also within the noun chunks. Presumably, I cant do it in the first step, i.e. before re-creating the `Doc` object, since I need to retain same sequence length right. Moreover, I would like entity labels to be consolidated into one label. Your example produces noun chunks with masked entities like so:
`CARDINAL CARDINAL CARDINAL movie, television and other film entertainment title`
Desired:
`CARDINAL movie, television and other film entertainment title`

**Reply by polm** [reply]
Can you give some more information about why you're trying to do this, or what your overall goal is?

> However, what I couldn't quite figure out yet: Where in the workflow would I remove stop words or punctuation?

For modern NLP techniques, removing stopwords and punctuation is not necessary, and will often have negative effects on model performance. It would help to know why you're trying to do that, and I would encourage you to reconsider whether it's necessary or a good idea.

If you want to do more complicated things like token merging it's possible in spaCy, but you'll have to fiddle with the Retokenizer and so on, and things will get complicated.

**Reply by simonschoe** [reply]
Sure. So I am employing a BoW approach where I'd like to compare longer texts based on their similarity. I employ `spacy` in a first step to pre-process the raw texts before I pass them on to `sklearn` for vectorization and so on. The text lengths (I am working with fairly long documents) as well as the research community in my field, prevents using contextualized embeddings stemming from `transformers`. However, I'd like to employ the `transformers` based pipeline to leverage the improved tagging and parsing capabilities.

That being said, I'd like to reduce the dimensionality of my BoW embeddings by removing stop words, punctuation, replace entities (especially numbers) by entity labels. Further, I'd like to at least account for noun chunks to partially account for local context which I would otherwise ignore following the BoW approach.

In essence: I largely have to stick to the more traditional NLP techniques, but I would like to apply these consistently, removing stop words and punctuation not only on the token-level but similarly within the noun chunks identified by the parser...

**Reply by polm** [reply]
OK, that makes sense, thanks for the explanation. 

To mask entities, you can use the `merge_entities` function in your initial pipeline to treat them as single tokens, and then handle them like I did `like_num` above. Example:

```
import spacy
from spacy.pipeline.functions import merge_noun_chunks
from spacy.tokens import Doc

nlp = spacy.load("en_core_web_sm")

nlp.add_pipe("merge_entities")

text = "more than 14,500 movie, television and other filmed entertainment titles"
doc = nlp(text)
for chunk in doc.noun_chunks:
    print(chunk)

print("-----")

def hide_num(tok):
    return "NUM" if tok.ent_type_ == "CARDINAL" else tok.text

pos = [tok.pos_ for tok in doc]
spaces = [bool(tok.whitespace_) for tok in doc]
words = [hide_num(tok) for tok in doc]
deps = [tok.dep_ for tok in doc]
heads = [tok.head.i for tok in doc]

doc2 = Doc(nlp.vocab, words=words, spaces=spaces, pos=pos, deps=deps, heads=heads)

for chunk in merge_noun_chunks(doc2).noun_chunks:
    print(chunk)
```

For stopwords and punctuation, if you're using noun chunks, I would just use noun chunks in place of the rest of the sentence rather than trying to filter stuff out. There's no good way to get noun chunks after removing stopwords and punctuation (nouns will run together and give false chunks or nonsense), and they'll probably be most of what's left. If you want you can separately get verbs and adverbs, and also predicate adjectives if you think they're common enough to be important.

That said, assuming you're using some kind of locality sensitive hashing with ngrams, I would recommend trying things without fancy preprocessing to start off. Specific numbers can often be a hint that documents started out the same, if you're looking for plagiarism/exact copy style similarity (as might be found in bits of long financial documents copied around), and more generally something like the number of digits can also be a hint.

**Reply by simonschoe** [reply]
Again, thank you so much for the help! I am satisfied with how entities are handled now and I think I'll proceed with the noun chunks as recommended by you for now (i.e. keeping punct and stops in those phrases).

To add some more context: The documents I am supposed to compare are published yearly and change little. Regularly, the authors only update the numbers inside, like adjusting years, monetary values, dates and so on and so forth. However, if they change I want to be sure to capture the change in language as accurate as possible. Therefore, I am not particularly interested in varying numbers as I don't want updated numbers to drive the similarity measure.

**Reply by polm** [reply]
Thanks for the context, that certainly explains why you'd want to ignore numbers. 

>  keeping punct and stops in those phrases

I don't think noun chunks will include any punctuation except serial commas, and the only stop words would be determiners (unless you consider "good" a stop word or something).
